### PR TITLE
Implement ignoring shebang on a tokenizer level

### LINKED ISSUE #81: Ignore shebang
Currently parsing files (with Selene) throw a parse error when a file has a shebang set.

```
#!/usr/bin/lua
```

### PR DESCRIPTION
**Pull request by zummenix**
Since the crate builds only with roblox feature, I've placed a test case in `roblox_cases`. It should be probably moved in general tests.

Closes #81

### CODE REVIEW

**Comment by Kampfkarren** [comment]
> Since the crate builds only with roblox feature, I've placed a test case in roblox_cases. 

I'm going to spend some time today getting it to work without the Roblox feature. 😄

**Comment by Kampfkarren** [comment]
Should be fixed on latest master.

**Comment by zummenix** [comment]
Done! 

I have a question about CI: do you deliberately don't test with `roblox` feature?

**Comment by Kampfkarren** [comment]
We test both with and without the feature flag.

File: full-moon/src/ast/owned.rs [hunk 1]
```diff
@@ -88,6 +88,9 @@ impl Owned for TokenType<'_> {
     fn owned(&self) -> Self::Owned {
         match self {
             TokenType::Eof => TokenType::Eof,
+            TokenType::Shebang { line } => TokenType::Shebang {
+                line: Cow::Owned(line.clone().into_owned()),
+            },
```
Comment (Kampfkarren) [inline full-moon/src/ast/owned.rs:93]: Move this to above SingleLineComment for alphabetical sorting.

File: full-moon/src/tokenizer.rs [hunk 2]
```diff
@@ -104,6 +104,13 @@ pub enum TokenType<'a> {
     /// End of file, should always be the very last token
     Eof,
 
+    /// A shebang line
+    Shebang {
+        #[cfg_attr(feature = "serde", serde(borrow))]
+        /// The shebang line itself
+        line: Cow<'a, str>,
+    },
+
```
Comment (Kampfkarren) [inline full-moon/src/tokenizer.rs:113]: Alphabetical order.

File: full-moon/src/tokenizer.rs [hunk 3]
```diff
@@ -200,6 +208,7 @@ impl<'a> TokenType<'a> {
     pub fn kind(&self) -> TokenKind {
         match self {
             TokenType::Eof => TokenKind::Eof,
+            TokenType::Shebang { .. } => TokenKind::Shebang,
```
Comment (Kampfkarren) [inline full-moon/src/tokenizer.rs:211]: Alphabetical order

File: full-moon/src/tokenizer.rs [hunk 4]
```diff
@@ -230,6 +239,8 @@ impl<'a> TokenType<'a> {
 pub enum TokenKind {
     /// End of file, should always be the very last token
     Eof,
+    /// The shebang line
+    Shebang,
```
Comment (Kampfkarren) [inline full-moon/src/tokenizer.rs:243]: Alphabetical order

File: full-moon/src/tokenizer.rs [hunk 5]
```diff
@@ -295,6 +306,7 @@ impl<'a> fmt::Display for Token<'a> {
 
         match &*self.token_type() {
             Eof => "".to_string(),
+            Shebang { line } => line.to_string(),
```
Comment (Kampfkarren) [inline full-moon/src/tokenizer.rs:309]: Alphabetical order

File: full-moon/src/tokenizer.rs [hunk 6]
```diff
@@ -364,6 +377,7 @@ impl<'ast> VisitMut<'ast> for Token<'ast> {
 
         match token.token_kind() {
             TokenKind::Eof => token,
+            TokenKind::Shebang => token,
```
Comment (Kampfkarren) [inline full-moon/src/tokenizer.rs:380]: Alphabetical order

File: full-moon/src/tokenizer.rs [hunk 7]
```diff
@@ -347,6 +359,7 @@ impl<'ast> Visit<'ast> for Token<'ast> {
 
         match self.token_kind() {
             TokenKind::Eof => {}
+            TokenKind::Shebang => {}
```
Comment (Kampfkarren) [inline full-moon/src/tokenizer.rs:362]: Alphabetical order

**Comment by Kampfkarren** [comment]
Can you add a test to fail cases for shebangs that are not at the beginning of the file?

**Comment by zummenix** [comment]
> Move this to above SingleLineComment for alphabetical sorting.

Ah, that's why `Eof` is the first 😄

**Comment by Kampfkarren** [comment]
One last request--can you add this to the CHANGELOG as well?

File: CHANGELOG.md [hunk 8]
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Added support for exporting types (`export type Foo = { bar: number }`) under the `roblox` feature flag
 - Added support for using types from other modules (`local x: module.Foo`) under the `roblox` feature flag
+- Added ability to ignore shebang
```
Comment (Kampfkarren) [inline CHANGELOG.md:11]:
```suggestion
- Added support for parsing a shebang
```

**Comment by Kampfkarren** [comment]
Thanks!

**Pull request merged**